src/ethereum/spurious_dragon/spec.py

```diff
@@ -0,0 +1,820 @@
+    transaction_fee = (tx.gas - gas_left - gas_refund) * tx.gas_price
+    total_gas_used = gas_used - gas_refund
+
+    # refund gas
+    sender_balance_after_refund = (
+        get_account(env.state, sender).balance + gas_refund_amount
+    )
+    set_account_balance(env.state, sender, sender_balance_after_refund)
+
+    # transfer miner fees
+    coinbase_balance_after_mining_fee = (
+        get_account(env.state, env.coinbase).balance + transaction_fee
+    )
+    set_account_balance(
+        env.state, env.coinbase, coinbase_balance_after_mining_fee
+    )
+
+    for address in accounts_to_delete:
+        destroy_account(env.state, address)
+
+    for address in touched_accounts:
+        should_delete = account_exists(
+            env.state, address
+        ) and is_account_empty(env.state, address)
+        if should_delete:
+            destroy_account(env.state, address)
+
+    return total_gas_used, logs
+
+
+def validate_transaction(tx: Transaction) -> bool:
+    """
+    Verifies a transaction.
+
+    Parameters
+    ----------
+    tx :
+        Transaction to validate.
+
+    Returns
+    -------
+    verified : `bool`
+        True if the transaction can be executed, or False otherwise.
+    """
+    return calculate_intrinsic_cost(tx) <= tx.gas and tx.nonce < 2 ** 64 - 1
+
+
+def calculate_intrinsic_cost(tx: Transaction) -> Uint:
+    """
+    Calculates the intrinsic cost of the transaction that is charged before
+    execution is instantiated.
+
+    Parameters
+    ----------
+    tx :
+        Transaction to compute the intrinsic cost of.
+
+    Returns
+    -------
+    verified : `eth1spec.base_types.Uint`
+        The intrinsic cost of the transaction.
+    """
+    data_cost = 0
+
+    for byte in tx.data:
+        if byte == 0:
+            data_cost += TX_DATA_COST_PER_ZERO
+        else:
+            data_cost += TX_DATA_COST_PER_NON_ZERO
+
+    if tx.to == Bytes0(b""):
+        create_cost = TX_CREATE_COST
+    else:
+        create_cost = 0
+
+    return Uint(TX_BASE_COST + data_cost + create_cost)
+
+
+def recover_sender(tx: Transaction) -> Address:
+    """
+    Extracts the sender address from a transaction.
+
+    Parameters
+    ----------
+    tx :
+        Transaction of interest.
+
+    Returns
+    -------
+    sender : `eth1spec.eth_types.Address`
+        The address of the account that signed the transaction.
+    """
+    v, r, s = tx.v, tx.r, tx.s
+
+    ensure(0 < r and r < SECP256K1N)
+    ensure(0 < s and s <= SECP256K1N // 2)
+
+    if v == 27 or v == 28:
+        public_key = crypto.secp256k1_recover(
+            r, s, v - 27, signing_hash_legacy(tx)
+        )
+    else:
+        ensure(v == 35 + CHAIN_ID * 2 or v == 36 + CHAIN_ID * 2)
+        public_key = crypto.secp256k1_recover(
+            r, s, v - 35 - CHAIN_ID * 2, signing_hash_155(tx)
+        )
+    return Address(crypto.keccak256(public_key)[12:32])
+
+
+def signing_hash_legacy(tx: Transaction) -> Hash32:
```

Comment:
I worry that calling something `legacy` might be... difficult over time, especially if a function is replaced multiple times.

In this particular instance, maybe `signing_hash_any_chain` and `signing_hash_single_chain`?

Reply:
Went for `signing_hash_pre155()`.